dask/layers.py

```diff
@@ -1239,3 +1262,341 @@ def __repr__(self):
+        the function specified by ``tree_node_func``. Note that the
+        first positional argument will always correspond to a list of
+        inputs from the input nodex (and these arguments will always
+        come after).
+    finalize_func : callable, optional
+        Function used in place of ``tree_node_func`` on the final tree
+        node(s) to produce the final output for each split. By default,
+        ``tree_node_func`` will be used.
+    finalize_kwargs : dict, optional
+        Dictionary of key-word arguments to include in every call to
+        the function specified by ``finalize_func``. This parameter
+        is ignored if ``finalize_func`` is not set.
+    split_every : int, optional
+        This argument specifies the maximum number of input nodes
+        inputs to be handled by any one task in the tree. Defaults to 32.
+    split_out : int, optional
+        This argument specifies the number of output nodes in the
+        reduction tree. If ``split_out`` is set to an integer >=1, the
+        input tasks must contain data that can be indexed by a ``getitem``
+        operation with a key in the range ``[0, split_out)``.
+    output_splits : list, optional
+        List of required output partitions. This parameter is used
+        internally by Dask for high-level culling.
+    tree_node_name : str, optional
+        Name to use for intermediate tree-node tasks.
+    """
+
+    def __init__(
+        self,
+        name,
+        name_input,
+        npartitions_input,
+        concat_func,
+        tree_node_func,
+        tree_node_kwargs=None,
+        finalize_func=None,
+        finalize_kwargs=None,
+        split_every=32,
+        split_out=None,
+        output_splits=None,
+        tree_node_name=None,
+        annotations=None,
+    ):
+        super().__init__(annotations=annotations)
+        self.name = name
+        self.name_input = name_input
+        self.npartitions_input = npartitions_input
+        self.concat_func = concat_func
+        self.tree_node_func = tree_node_func
+        self.tree_node_kwargs = tree_node_kwargs or {}
+        self.finalize_func = finalize_func
+        self.finalize_kwargs = finalize_kwargs or {}
+        self.split_every = split_every
+        self.split_out = split_out
+        self.output_splits = output_splits
+        self.tree_node_name = tree_node_name or "tree_node-" + self.name
+
+        # Calculate tree withs and height
+        # (Used to get output keys without materializing)
+        parts = self.npartitions_input
+        self.widths = [parts]
+        while parts > 1:
+            parts = math.ceil(parts / self.split_every)
+            self.widths.append(parts)
+        self.height = len(self.widths)
+
+        # Set default output_splits
+        if self.output_splits is None:
+            self.output_splits = list(range(self.split_out or 1))
+
+    def _check_key(self, *task):
```

Comment:
```suggestion
    def _make_key(self, *name_parts, split_i):
```
Doesn't feel to me like this function is checking anything; it seems to be generating keys. And it doesn't take a task?

Reply:
Right, exactly - It's just checking if the input task key needs to have the last (split) element, and removing it if not. This was just to simplify the code a bit while I was working out the logic.  Going in the other direction (`_make_key`) certainly makes more sense in hindsight.